test/e2e/tests/vault-corruption/vault-corruption.spec.ts

```diff
@@ -0,0 +1,340 @@
+    });
+  }
+
+  /**
+   * Click the recovery/reset button then confirm or dismiss the action.
+   *
+   * @param driver - The WebDriver instance.
+   * @param confirm - Whether to confirm the action or not.
+   */
+  async function clickRecover(driver: Driver, confirm: boolean) {
+    // click the Recovery/Reset button
+    const recoveryButton = await getVaultRecoveryButton(driver);
+    await recoveryButton.click();
+
+    // Confirm we want to recover/reset.
+    const prompt = await driver.driver.switchTo().alert();
+    if (confirm) {
+      await prompt.accept();
+    } else {
+      await prompt.dismiss();
+    }
+
+    // the button should be disabled while the recovery process is in progress,
+    // and enabled if the user dismissed the prompt
+    const isEnabled = await recoveryButton.isEnabled();
+    assert.equal(
+      isEnabled,
+      !confirm,
+      confirm
+        ? 'Recovery button should be disabled when prompt is accepted'
+        : 'Recovery button should be enabled when prompt is dismissed',
+    );
+  }
+
+  /**
+   * Recovers the vault and returns the first account's address.
+   *
+   * @param driver - The WebDriver instance.
+   */
+  async function onboardAfterRecovery(driver: Driver) {
+    // Log back in to the wallet and complete onboarding
+    await completeVaultRecoveryOnboardingFlow({
+      driver,
+      password: WALLET_PASSWORD,
+    });
+
+    // now that we are re-onboarded, get the first account's address
+    return await getFirstAddress(driver);
+  }
+
+  /**
+   * Returns the first (truncated) address from the account list in UI.
+   *
+   * @param driver - The WebDriver instance.
+   * @param headerNavbar
+   */
+  async function getFirstAddress(
+    driver: Driver,
+    headerNavbar: HeaderNavbar = new HeaderNavbar(driver),
+  ) {
+    await headerNavbar.openAccountMenu();
+
+    const accountListPage = new AccountListPage(driver);
+    await accountListPage.check_pageIsLoaded();
+    await accountListPage.openAccountDetailsModal('Account 1');
+
+    const accountDetailsModal = new AccountDetailsModal(driver);
+    await accountDetailsModal.check_pageIsLoaded();
+
+    const accountAddress = await accountDetailsModal.getAccountAddress();
+    return accountAddress;
+  }
+
+  it('recovers metamask vault when primary database is broken but backup is intact', async function () {
+    await withFixtures(
+      getConfig(this.test?.title),
+      async ({ driver }: { driver: Driver }) => {
+        const initialFirstAddress = await onboardThenCorruptVault(
+          driver,
+          breakPrimaryDatabaseOnlyScript,
+        );
+
+        // start recovery
+        await clickRecover(driver, true);
+
+        // onboard again
+        const restoredFirstAddress = await onboardAfterRecovery(driver);
+
+        // make sure the address is the same as before
+        assert.equal(
+          restoredFirstAddress,
+          initialFirstAddress,
+          'Addresses should match',
```

Comment:
aww okay this type of assertion can also lead to race conditions, where the element is loaded in the DOM, but it yet not has the desired value. Then the assertion fails ([see](https://github.com/MetaMask/metamask-extension/issues/19870)).

I see that the `accountDetailsModal.getAccountAddress();` method in the page object class,  can lead to make this type of assertions. Given that this was already an existing method, I would leave this out of scope of this PR, but I'll create a ticket for us to fix that (I've seen this happens in a couple of more specs). To mitigate this type of race condition, instead of getting an address and asserting its value, we should look directly for the address we want

Reply:
`getAccountAddress` not always getting the account address is a bit unexpected. haha

I'll leave the fix out of this PR, just to make sure it does get fixed at some point. haha

I actually had a custom script to to this that did always work, but then I discovered the `getAccountAddress` function and thought it'd be better to just use that! Hahaha

I also noticed that `getAccountAddress` doesn't even return the full account address, it returns the _shortened_ address. It should probably return the whole thing.

Reply:
ohh interesting, I didn't notice the shortened address! Indeed I think that method needs some changes. It can be out of scope of the PR. I'll create a ticket and try to tackle this in the upcoming week :pray:

Reply:
issue created [here](https://github.com/MetaMask/metamask-extension/issues/33357)